packages/manager/modules/anthos/src/dashboard/components/assign-private-ip/assign-private-ip.controller.js

```diff
@@ -0,0 +1,92 @@
+import { IPV4_REGEX } from './constant';
+
+export default class AssignPrivateIpController {
+  /* @ngInject */
+  constructor($translate, Alerter, AnthosTenantsService) {
+    this.$translate = $translate;
+    this.Alerter = Alerter;
+    this.AnthosTenantsService = AnthosTenantsService;
+
+    this.IPV4_REGEX = IPV4_REGEX;
+    this.ipModel = {
+      form: null,
+      labels: {
+        privateInputIp: 'privateInputIp',
+      },
+      validationKeys: {
+        CIDR_IP: 'CIDR_IP',
+        PRIVATE_IP: 'PRIVATE_IP',
+      },
+      privateIp: {
+        range: null,
+      },
+    };
+  }
+
+  isValidCidrIp() {
+    return IPV4_REGEX.CIDR_RANGE_32.test(this.ipModel.privateIp.range);
+  }
+
+  isPrivateIp() {
+    return IPV4_REGEX.PRIVATE_IP.test(this.ipModel.privateIp.range);
+  }
+
+  isValidIp() {
+    const { form, labels } = this.ipModel;
+    const { CIDR_IP, PRIVATE_IP } = this.ipModel.validationKeys;
+
+    if (!this.isValidCidrIp()) {
+      form[labels.privateInputIp].$setValidity(CIDR_IP, false);
```

Comment:
Why can't we use ng-pattern for validation?

Reply:
Because we have to test different cases: private ip, cidr ip, range ip